<b>TP : Partons faire un peu de shopping ! Analyse d'un listing de 10 000 chaussures vendues en ligne</b>

source du dataset : https://www.kaggle.com/datafiniti/womens-shoes-prices

Nous allons analyser un data set contenant 10 000 chaussures avec de nombreuse caractéristiques comme le prix, les couleurs disponibles, les tailles disponibles, leurs poids...

Voici une traduction de la description des colonnes principales : 
- asins : Une liste des ASIN (identificateurs d'Amazon) utilisés pour ce produit.
- brand : Le nom de marque de ce produit.
- categories : Une liste de mots-clés de catégorie utilisés pour ce produit dans plusieurs sources.
- colors : Une liste de couleurs disponibles pour ce produit.
- count : Le nombre d'unités incluses dans l'emballage du produit. Peut inclure une description de l'unité.
- dateAdded : La date à laquelle ce produit a été ajouté pour la première fois à la base de données des produits.
- dateUpdated : La date la plus récente à laquelle ce produit a été mis à jour ou vu par notre système.
- descriptions : Une liste des descriptions de ce produit provenant de diverses sources. Chaque description contient :
- dimension : La longueur, la largeur et la hauteur de ce produit. Unités incluses.
- ean : Les codes EAN de ce produit. Il peut y avoir plus d'un EAN dans la liste lorsqu'un produit a plusieurs variantes,
- financingAndLeasing : Une liste des conditions de financement ou de location associées à ce produit
- features : Une liste des caractéristiques associées à ce produit. La liste des caractéristiques peut être obtenue à partir des sections de caractéristiques spécifiques disponibles dans la liste des produits du produit.
- manufacturer : Numéro du fabricant
- merchants : Une liste des marchands qui vendent ce produit. Il s'agit généralement de marchands tiers que l'on trouve sur les sites Web de commerce électronique.
- name : dénomination du produit
- prices : Une liste de prix pour ce produit.
- primaryCategories : Une liste des catégories normalisées auxquelles ce produit appartient. Les valeurs des catégories peuvent être trouvées ici.
- quantities : Une liste des quantités disponibles pour ce produit. Les quantités indiquent combien d'unités du produit sont disponibles auprès d'une source spécifique à un moment donné.
- reviews : Une liste de commentaires pour ce produit.
- sizes : Une liste des tailles disponibles pour ce produit.
- skus : Une liste des références de ce produit. Les SKU sont typiquement spécifiques aux détaillants individuels ou aux sites Web.
- sourceURLs : Une liste d'URLs utilisées pour générer les données de ce produit.
- upc : Le code CUP de ce produit.
- websiteIDs : Une liste d'identifiants de site Web pour ce produit. Les ID de site Web sont liés à un détaillant en ligne spécifique. Il s'agit de l'identifiant unique utilisé par le site Web. Ils sont utilisés par Datafiniti lorsque d'autres identificateurs uniques ne sont pas disponibles pour un produit.
- weight : Le poids du produit. Unités incluses.

La source https://developer.datafiniti.co/docs/product-data-schema

<b> Exercices </b>

<b>Découverte du data set</b><br>
1) à l'aide de la fonction read_csv, importez entièrement le data set de enregistrez dans le dataframe 'Shoes' <br>
2) Utilisez la méthode .head() pour afficher les premières lignes <br>
3) Il y a trop de colonnes...Pour avoir une vue d'ensemble changez le paramétrage de pandas avec set_option pour afficher les 34 colonnes. Une fois cela fait, affichez à nouveau le head de la page <br>
4) On va garder uniquement les colonnes qui nous interessent. grace à la méthode de sélection de plusieurs colonnes, créez un dataframe 'ShoesLight', comprenant unique les colonnes suivantes :'id','name','dateUpdated','colors','prices.amountMax','prices.amountMin','prices.merchant'. <br>Affichez le head de ShoesLight.<br>

<b> data cleaning </b> <br>
5) A l'aide de la méthode .dtypes, consultez les types de chaque colonne. Ecrivez en markdown celles devant être corrigées et comment. <br>
6) A l'aide des méthodes .isnull, .sum() et len(), calculez le % de valeurs maquantes par colonne. Notez en markdown celles qui sont problématiques. <br>
7) Il manque trop de valeurs pour les colonnes des e-commercant et des couleurs. Retirez les du dataframe shoesLight <br>
8) A l'aide de la méthode .to_datetime, transformez la colonne 'dateUpdated'

<b> features modeling </b> <br>
9) caculez la colonne 'prices.amountAverage' en additionnant les colonnes 'prices.amountMax' et 'prices.amountMin' et en divisant par 2 <br>
10) grace à la méthode dt.weekday, créez une colonne 'dayOfweekUpdated', permettant de connaitre à partir de dateUpdated le jour de la semaine où les produits sont mis à jour

<b> data analyse </b> <br>

11) Analyser le prix moyen des chaussures avec la methode .desribe() <br>
12) Y-a-t-il de grandes différences de prix en fonction de la marque ? A l'aide des methodes groupby(), mean() et sort_values(), créez une varaible 'Luxe' contenant les 10 marques les plus chères, puis la variable 'lowCost' contenant les 10 marques les moins chères. <br>
13) grace à la méthode value_counts(), determinez le jour de la semaine où les produits sont le plus souvent mis à jour <br>
14) (optionnel) donnez le prix moyen des produits de la marque 'easy street' mis à jour un jeudi


In [2]:
import pandas as pd
import numpy as np
Shoes=pd.read_csv('Datafiniti_Womens_Shoes.csv')

In [3]:
Shoes.head()

id             dateAdded           dateUpdated asins  \
0  AVpfEf_hLJeJML431ueH  2015-05-04T12:13:08Z  2018-01-29T04:38:43Z   NaN   
1  AVpi74XfLJeJML43qZAc  2017-01-27T01:23:39Z  2018-01-03T05:21:54Z   NaN   
2  AVpi74XfLJeJML43qZAc  2017-01-27T01:23:39Z  2018-01-03T05:21:54Z   NaN   
3  AVpjXyCc1cnluZ0-V-Gj  2017-01-27T01:25:56Z  2018-01-04T11:52:35Z   NaN   
4  AVphGKLPilAPnD_x1Nrm  2017-01-27T01:25:56Z  2018-01-18T03:55:18Z   NaN   

         brand                                         categories  \
0  Naturalizer  Clothing,Shoes,Women's Shoes,All Women's Shoes...   
1     MUK LUKS  Clothing,Shoes,Women's Shoes,Women's Casual Sh...   
2     MUK LUKS  Clothing,Shoes,Women's Shoes,Women's Casual Sh...   
3     MUK LUKS  Clothing,Shoes,Women's Shoes,All Women's Shoes...   
4     MUK LUKS  Clothing,Shoes,Women's Shoes,All Women's Shoes...   

  primaryCategories                          colors                 dimension  \
0             Shoes  Silver,Cream Watercolor Floral                       NaN   
1             Shoes                            Grey                       NaN   
2             Shoes                            Grey                       NaN   
3       Shoes,Shoes                           Black  6.0 in x 6.0 in x 1.0 in   
4             Shoes                            Grey  6.0 in x 6.0 in x 1.0 in   

            ean  ...   prices.merchant prices.offer prices.returnPolicy  \
0           NaN  ...     Overstock.com          NaN                 NaN   
1  3.397705e+10  ...       Walmart.com          NaN                 NaN   
2  3.397705e+10  ...  Slippers Dot Com          NaN                 NaN   
3  3.397705e+10  ...  Slippers Dot Com          NaN                 NaN   
4  3.397705e+10  ...       Walmart.com          NaN                 NaN   

  prices.shipping prices.size  \
0             NaN           S   
1        Standard           6   
2           Value           6   
3           Value           6   
4       Expedited           6   

                                   prices.sourceURLs  \
0  https://www.overstock.com/Clothing-Shoes/Women...   
1  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...   
2  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...   
3  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...   
4  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...   

                                               sizes  \
0  6W,9W,7.5W,12W,8.5M,9N,9M,9.5M,10.5M,10W,8.5W,...   
1                                         10,7,6,9,8   
2                                         10,7,6,9,8   
3                                         10,7,6,9,8   
4                                         10,7,6,9,8   

                                          sourceURLs           upc weight  
0  https://www.walmart.com/ip/Naturalizer-Danya-W...  017136472311    NaN  
1  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...  033977045743    NaN  
2  https://www.walmart.com/ip/MUK-LUKS-Womens-Jan...  033977045743    NaN  
3  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...  033977045903    NaN  
4  https://www.walmart.com/ip/MUK-LUKS-Womens-Daw...  033977045958    NaN  

[5 rows x 34 columns]

In [4]:
pd.set_option("display.max_columns",34) # Pour éviter les 3points et afficher le nombre maximal(34) de colonnes qu'on souhaite afficher

In [5]:
Shoes.head()

id             dateAdded           dateUpdated asins  \
0  AVpfEf_hLJeJML431ueH  2015-05-04T12:13:08Z  2018-01-29T04:38:43Z   NaN   
1  AVpi74XfLJeJML43qZAc  2017-01-27T01:23:39Z  2018-01-03T05:21:54Z   NaN   
2  AVpi74XfLJeJML43qZAc  2017-01-27T01:23:39Z  2018-01-03T05:21:54Z   NaN   
3  AVpjXyCc1cnluZ0-V-Gj  2017-01-27T01:25:56Z  2018-01-04T11:52:35Z   NaN   
4  AVphGKLPilAPnD_x1Nrm  2017-01-27T01:25:56Z  2018-01-18T03:55:18Z   NaN   

         brand                                         categories  \
0  Naturalizer  Clothing,Shoes,Women's Shoes,All Women's Shoes...   
1     MUK LUKS  Clothing,Shoes,Women's Shoes,Women's Casual Sh...   
2     MUK LUKS  Clothing,Shoes,Women's Shoes,Women's Casual Sh...   
3     MUK LUKS  Clothing,Shoes,Women's Shoes,All Women's Shoes...   
4     MUK LUKS  Clothing,Shoes,Women's Shoes,All Women's Shoes...   

  primaryCategories                          colors                 dimension  \
0             Shoes  Silver,Cream Watercolor Floral                       NaN   
1             Shoes                            Grey                       NaN   
2             Shoes                            Grey                       NaN   
3       Shoes,Shoes                           Black  6.0 in x 6.0 in x 1.0 in   
4             Shoes                            Grey  6.0 in x 6.0 in x 1.0 in   

            ean                                          imageURLs  \
0           NaN  https://i5.walmartimages.com/asr/861ca6cf-fa55...   
1  3.397705e+10  https://i5.walmartimages.com/asr/421de5d5-3a74...   
2  3.397705e+10  https://i5.walmartimages.com/asr/421de5d5-3a74...   
3  3.397705e+10  https://i5.walmartimages.com/asr/950d38a5-0113...   
4  3.397705e+10  https://i5.walmartimages.com/asr/5e137bc3-c900...   

                                                keys manufacturer  \
0  naturalizer/47147sc022,017136472311,womensnatu...          NaN   
1  mukluks/00173650206,033977045743,muklukswomens...     Muk Luks   
2  mukluks/00173650206,033977045743,muklukswomens...     Muk Luks   
3  033977045903,muklukswomensdawnsuedescuffslippe...     Muk Luks   
4  mukluks/00173660206,033977045958,0033977045958...          NaN   

  manufacturerNumber                                               name  \
0         47147SC022  Naturalizer Danya Women N/S Open Toe Synthetic...   
1       0017365020-6                MUK LUKS Womens Jane Suede Moccasin   
2       0017365020-6                MUK LUKS Womens Jane Suede Moccasin   
3       0017366001-6           MUK LUKS Womens Dawn Suede Scuff Slipper   
4       0017366020-6           MUK LUKS Womens Dawn Suede Scuff Slipper   

   prices.amountMax  prices.amountMin prices.availability        prices.color  \
0             55.99             55.99                 NaN  UWomens  M Regular   
1             47.00             35.25            In Stock                Grey   
2             35.25             35.25            In Stock                Grey   
3             24.75             24.75            In Stock               Black   
4             33.00             30.39            In Stock                Grey   

  prices.condition prices.currency      prices.dateAdded  \
0              NaN             USD  2017-03-28T11:40:25Z   
1              New             USD  2018-01-03T05:21:54Z   
2              New             USD  2017-12-06T05:02:42Z   
3              New             USD  2018-01-04T11:52:35Z   
4              New             USD  2017-12-04T21:35:47Z   

                                     prices.dateSeen  prices.isSale  \
0  2017-03-25T09:19:24.819Z,2017-03-25T09:19:19.600Z          False   
1  2017-12-08T14:24:00.000Z,2017-11-01T02:52:00.000Z           True   
2  2017-11-10T15:11:00.000Z,2017-11-18T08:00:00.000Z          False   
3                           2017-12-07T16:37:00.000Z          False   
4                           2017-11-17T21:15:00.000Z           True   

    prices.merchant prices.offer  prices.returnPolicy prices.shipping  \
0     Overstock.com  

In [7]:
ShoesLight=Shoes[['id','name','dateUpdated','colors','prices.amountMax','prices.amountMin','prices.merchant']]
ShoesLight.head()

id                                               name  \
0  AVpfEf_hLJeJML431ueH  Naturalizer Danya Women N/S Open Toe Synthetic...   
1  AVpi74XfLJeJML43qZAc                MUK LUKS Womens Jane Suede Moccasin   
2  AVpi74XfLJeJML43qZAc                MUK LUKS Womens Jane Suede Moccasin   
3  AVpjXyCc1cnluZ0-V-Gj           MUK LUKS Womens Dawn Suede Scuff Slipper   
4  AVphGKLPilAPnD_x1Nrm           MUK LUKS Womens Dawn Suede Scuff Slipper   

            dateUpdated                          colors  prices.amountMax  \
0  2018-01-29T04:38:43Z  Silver,Cream Watercolor Floral             55.99   
1  2018-01-03T05:21:54Z                            Grey             47.00   
2  2018-01-03T05:21:54Z                            Grey             35.25   
3  2018-01-04T11:52:35Z                           Black             24.75   
4  2018-01-18T03:55:18Z                            Grey             33.00   

   prices.amountMin   prices.merchant  
0             55.99     Overstock.com  
1             35.25       Walmart.com  
2             35.25  Slippers Dot Com  
3             24.75  Slippers Dot Com  
4             30.39       Walmart.com

In [8]:
ShoesLight.dtypes # Affichage du type de chaque colonne

id                   object
name                 object
dateUpdated          object
colors               object
prices.amountMax    float64
prices.amountMin    float64
prices.merchant      object
dtype: object

In [9]:
ShoesLight.isnull().sum()/len(ShoesLight)

id                  0.0000
name                0.0000
dateUpdated         0.0000
colors              0.7369
prices.amountMax    0.0000
prices.amountMin    0.0000
prices.merchant     0.9565
dtype: float64

In [10]:
ShoesLight=ShoesLight[['id','name','dateUpdated','prices.amountMax','prices.amountMin']]

In [11]:
ShoesLight['dateUpdated']=pd.to_datetime(ShoesLight['dateUpdated']) # Changement de type en datetime
ShoesLight.head()

id                                               name  \
0  AVpfEf_hLJeJML431ueH  Naturalizer Danya Women N/S Open Toe Synthetic...   
1  AVpi74XfLJeJML43qZAc                MUK LUKS Womens Jane Suede Moccasin   
2  AVpi74XfLJeJML43qZAc                MUK LUKS Womens Jane Suede Moccasin   
3  AVpjXyCc1cnluZ0-V-Gj           MUK LUKS Womens Dawn Suede Scuff Slipper   
4  AVphGKLPilAPnD_x1Nrm           MUK LUKS Womens Dawn Suede Scuff Slipper   

                dateUpdated  prices.amountMax  prices.amountMin  
0 2018-01-29 04:38:43+00:00             55.99             55.99  
1 2018-01-03 05:21:54+00:00             47.00             35.25  
2 2018-01-03 05:21:54+00:00             35.25             35.25  
3 2018-01-04 11:52:35+00:00             24.75             24.75  
4 2018-01-18 03:55:18+00:00             33.00             30.39

In [19]:
ShoesLight['prices.amountAverage']=(ShoesLight['prices.amountMax'] + ShoesLight['prices.amountMin'])/2
ShoesLight.head()

id                                               name  \
0  AVpfEf_hLJeJML431ueH  Naturalizer Danya Women N/S Open Toe Synthetic...   
1  AVpi74XfLJeJML43qZAc                MUK LUKS Womens Jane Suede Moccasin   
2  AVpi74XfLJeJML43qZAc                MUK LUKS Womens Jane Suede Moccasin   
3  AVpjXyCc1cnluZ0-V-Gj           MUK LUKS Womens Dawn Suede Scuff Slipper   
4  AVphGKLPilAPnD_x1Nrm           MUK LUKS Womens Dawn Suede Scuff Slipper   

                dateUpdated  prices.amountMax  prices.amountMin  \
0 2018-01-29 04:38:43+00:00             55.99             55.99   
1 2018-01-03 05:21:54+00:00             47.00             35.25   
2 2018-01-03 05:21:54+00:00             35.25             35.25   
3 2018-01-04 11:52:35+00:00             24.75             24.75   
4 2018-01-18 03:55:18+00:00             33.00             30.39   

   prices.amountAverage  
0                55.990  
1                41.125  
2                35.250  
3                24.750  
4                31.695

In [20]:
ShoesLight.describe()

prices.amountMax  prices.amountMin  prices.amountAverage
count      10000.000000      10000.000000          10000.000000
mean          69.223544         51.131209             60.177377
std           19.487286         21.267446             18.786193
min            5.870000          4.880000              5.375000
25%           59.990000         37.490000             49.990000
50%           64.990000         49.990000             56.875000
75%           79.990000         59.990000             67.490000
max          359.950000        359.950000            359.950000

Il y'a une grande différence de prix car le prix maximal est de 359.95 alors que le prix minimal est de 5.375

In [21]:
luxe=ShoesLight[['name','prices.amountAverage']].groupby(['name']).mean().sort_values(['prices.amountAverage'],ascending=False).head(10)
luxe

prices.amountAverage
name                                                                    
Red Wing Harriet Boots - Women's                                  359.95
Red Wing 6-inch Classic Moc Boots - Women's                       289.95
Asics GEL-FujiEndurance Trail Running Shoe - Wo...                179.95
Free People Los Valley Ankle Boot - Women's                       178.00
La Sportiva Ultra Raptor GTX Trail Running Shoe...                165.00
La Sportiva Akyra GTX Shoe - Women's                              160.00
On Footwear Cloudflyer Running Shoe - Women's                     159.99
Hoka One One Stinson ATR 4 Trail Running Shoe -...                159.95
Hoka One One Elevon Running Shoe - Women's                        159.95
New Balance Fuel Cell v1 Running Shoe - Women's                   159.95

In [22]:
lowCost=ShoesLight[['name','prices.amountAverage']].groupby(['name']).mean().sort_values(['prices.amountAverage'],ascending=False).tail(10)
lowCost

prices.amountAverage
name                                                                    
Apt. 9 Luckily Women's Sandals                                 16.859048
Earth Spirit Women's Andi Sandal                               16.380000
Victoria K. Women's Fashion Chain Link Strap Sa...             15.880000
LC Lauren Conrad Pixii Women's Flip Flops                      15.745556
Women's MUK LUKS® Soft Ones Sock Money Slippers               10.950000
Faded Glory Women's Printed Twin Gore Canvas Shoe              10.160000
Women's Essential Light Weight Athletic Shoe                    9.620000
Faded Glory Women's Casual Lace-up Canvas Sneaker               8.970000
Women's Essential Ballet Flat                                   5.415000
Time and Tru Women's Basic Ballet Flat                          5.375000